### PR TITLE
Add Katib Client in v1alpha2

### PR DESCRIPTION
**Pull request by andreyvelich**
I added Katib Client functionality in the util.
We need it for the UI, like we have here: 
https://github.com/kubeflow/katib/blob/master/pkg/manager/v1alpha1/studyjobclient/studyjobclient.go.

Will we have config map with Trial templates and Metrics collector templates, like here: 
https://github.com/kubeflow/katib/blob/master/manifests/v1alpha1/studyjobcontroller/metricsControllerConfigMap.yaml
https://github.com/kubeflow/katib/blob/master/manifests/v1alpha1/studyjobcontroller/workerConfigMap.yaml ?
I named these config maps as:
`trial-template` and  `metricscollector-template`

/assign @hougangliu @YujiOshima @johnugeorge 
/cc @richardsliu

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/480)

### CODE REVIEW

File: pkg/controller/v1alpha2/experiment/util/client_util.go [hunk 1]
```diff
@@ -16,14 +16,118 @@ limitations under the License.
...
 )
 
-func NewClient(options client.Options) (client.Client, error) {
+type katibClient struct {
+	client client.Client
+}
+
+const (
+	trialTemplatesName            = "trial-template"
+	metricsCollectorTemplatesName = "metricscollector-template"
```
Comment (johnugeorge) [inline pkg/controller/v1alpha2/experiment/util/client_util.go]: How about adding it in pkg/api/operators/apis/experiment/v1alpha2/constants.go?  trialTemplate is already defined

File: pkg/controller/v1alpha2/experiment/util/client_util.go [hunk 2]
```diff
@@ -16,14 +16,118 @@ limitations under the License.
...
 import (
+	"context"
+	"io/ioutil"
+	"strings"
+
+	experimentv1alpha2 "github.com/kubeflow/katib/pkg/api/operators/apis/experiment/v1alpha2"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	cl "sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (hougangliu) [inline pkg/controller/v1alpha2/experiment/util/client_util.go]: duplicate with line 26

File: pkg/controller/v1alpha2/experiment/util/client_util.go [hunk 3]
```diff
@@ -16,14 +16,118 @@ limitations under the License.
...
+		return err
+	}
+	return nil
+}
+
+func (k *katibClient) GetTrialTemplates(namespace ...string) (map[string]string, error) {
+
+	ns := getNamespace(namespace...)
+	trialTemplates := &apiv1.ConfigMap{}
+
```
Comment (hougangliu) [inline pkg/controller/v1alpha2/experiment/util/client_util.go]: Suggest move getConfigMap into this file, too and then please call getConfigMap

File: pkg/controller/v1alpha2/experiment/util/client_util.go [hunk 4]
```diff
@@ -16,14 +16,118 @@ limitations under the License.
...
+func (k *katibClient) CreateExperiment(experiment *experimentv1alpha2.Experiment, namespace ...string) error {
+
+	if err := k.client.Create(context.Background(), experiment); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *katibClient) GetTrialTemplates(namespace ...string) (map[string]string, error) {
```
Comment (hougangliu) [inline pkg/controller/v1alpha2/experiment/util/client_util.go]: for now, trial template can be stored in any configmap, not just `trial-template`

**Comment by richardsliu** [comment]
@andreyvelich Most likely yes, we will have similar metrics collector templates.

Can we rename the template as `metrics-collector-template`?

File: pkg/controller/v1alpha2/experiment/util/client_util.go [hunk 5]
```diff
@@ -16,14 +16,118 @@ limitations under the License.
...
+	trialTemplates.Data = newTrialTemplates
+
+	if err := k.client.Update(context.Background(), trialTemplates); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *katibClient) GetMetricsCollectorTemplates(namespace ...string) (map[string]string, error) {
+
```
Comment (hougangliu) [inline pkg/controller/v1alpha2/experiment/util/client_util.go]: I wonder if we have MetricsCollectorTemplates in v1alpha2

File: pkg/controller/v1alpha2/experiment/util/client_util.go [hunk 6]
```diff
@@ -16,14 +16,118 @@ limitations under the License.
...
+	if err := k.client.Get(context.Background(), types.NamespacedName{Name: metricsCollectorTemplatesName, Namespace: ns}, metricsCollectorTemplates); err != nil {
+		return nil, err
+	}
+	return metricsCollectorTemplates.Data, nil
+
+}
+
+func (k *katibClient) UpdateMetricsCollectorTemplates(newMCTemplates map[string]string, namespace ...string) error {
+
```
Comment (hougangliu) [inline pkg/controller/v1alpha2/experiment/util/client_util.go]: ditto

**Comment by johnugeorge** [comment]
Client code can be common util code which can be used by all components.  I think, it should be moved out of experiment controller

**Comment by richardsliu** [comment]
We can create a `pkg/util/v1alpha2` directory for common clients.

**Comment by andreyvelich** [comment]
@johnugeorge @richardsliu What do you think about moving it into manager folder, like here: 
https://github.com/kubeflow/katib/tree/master/pkg/manager/v1alpha1/studyjobclient

File: pkg/controller/v1alpha2/experiment/util/client_util.go (hunk 4 above)
Comment (andreyvelich) [reply pkg/controller/v1alpha2/experiment/util/client_util.go]: Is there is any way to get Trial templates configmap?

File: pkg/controller/v1alpha2/experiment/util/client_util.go (hunk 3 above)
Comment (andreyvelich) [reply pkg/controller/v1alpha2/experiment/util/client_util.go]: Done

**Comment by richardsliu** [comment]
@andreyvelich I think we need something like that, but I am not sure why that directory is named "manager". For v1alpha1 it just contains:
- metricscollector
- file-metricscollector
- studyjobclient
- visualize

Seems more like an util folder to me.

**Comment by andreyvelich** [comment]
@richardsliu Should we name this folder util in v1alpha2 ?

**Comment by richardsliu** [comment]
@andreyvelich Sounds good.

**Comment by andreyvelich** [comment]
@richardsliu @johnugeorge Done

File: pkg/api/operators/apis/experiment/v1alpha2/constants.go [hunk 7]
```diff
@@ -27,4 +27,10 @@ const (
 
 	// Default value of Spec.TemplatePath
 	DefaultTrialTemplatePath = "defaultTrialTemplate.yaml"
+
+	// Name of the configMap for Trial templates
+	TrialTemplatesName = "trial-template"
```
Comment (johnugeorge) [inline pkg/api/operators/apis/experiment/v1alpha2/constants.go]: line no 23 has it added already

File: pkg/api/operators/apis/experiment/v1alpha2/constants.go [hunk 8]
```diff
@@ -27,4 +27,10 @@ const (
...
+
+	// Name of the configMap for MetricsCollector templates
+	MetricsCollectorTemplatesName = "metrics-collector-template"
```
Comment (johnugeorge) [inline pkg/api/operators/apis/experiment/v1alpha2/constants.go]: Is this the default metric config map similar to trial?

Comment (andreyvelich) [reply pkg/api/operators/apis/experiment/v1alpha2/constants.go]: Yes, I will rename it, like for Trial template

**Pull request merged**